### DISCUSSION TITLE
Cross Probing Model Clipping - Emulation in different screens

### DISCUSSION THREAD

**Discussion by tonygem2022** [iTwin/itwinjs-core#4707] [General]
Hi iTwin Dev team.

We have deployed the cross probing example where we use a reality model and a CAD model in split screen mode.

Whilst we use clipping in the reality model screen and it works as intended, is it possible to have the clipping emulated in the second CAD model screen automatically with the identical behaviours and extents? (and vice versa - i.e. clipping in CAD model screen and automatically emulating in reality model) 

Thanks.

**Comment by pmconne** [comment]
Hi @tonygem2022,

[This](https://www.itwinjs.org/sandboxes/iTwinPlatform/Cross-Probing) is the cross-probing sample, for correlating 2d and 3d representations of objects in two viewports.
[This](https://www.itwinjs.org/sandboxes/iTwinPlatform/Swiping%20Comparison) is the swiping comparison sample, for displaying split-screen viewports - one showing a reality model, the other a design model, with clipping and synchronized frustums.

Can you please be more specific about what you're trying to achieve?

Thanks,
Paul

**Comment by tonygem2022** [comment]
Thanks for the timely revert @pmconne 

We have deployed the first sample ok. Our focus is on reality model and CAD model cross probing.

We use the element select tool and then use the clipping features to isolate elements in the reality model. When we do this, we want the same action to be reflected automatically in the CAD cross probing environment.

Whilst the "click and zoom to" capability works, we want the clipping features to work as well.

Hope this articulate the question better.

Because of the sensitive nature of the data set, i am unable to attach a screen shot but can send something privately if this helps.

Thanks.

**Reply by pmconne** [reply]
You can synchronize the states of two or more viewports using [connectViewports](https://www.itwinjs.org/reference/core-frontend/views/connectviewports/).
If that doesn't help, can you show the code you're currently using to synchronize the viewports?

**Comment by tonygem2022** [comment]
Hello @pmconne et al.

Refer to the image below please.

We have cross probing working between a reality view and a CAD model. We have connected viewports.

What we want to achieve is use the clipping features (drag arrows) to isolate elements in the reality model. When we do this, we want the same action to be reflected automatically in the CAD cross probing environment (and vis-a-vis).

Currently we can only clip in the reality model and the change in the box is not dynamically performing the same action in the CAD model and only showing the same volume in the CAD model as it does in the Reality view.

![Clip Issue](https://user-images.githubusercontent.com/108276006/204445653-7902a44f-6084-49f7-980f-783adc6acf30.png)

Please advise how this can be achieved.

Thanks.

**Reply by pmconne** [reply]
> can you show the code you're currently using to synchronize the viewports?

**Reply by pmconne** [reply]
I don't know why I can't see your reply (did you delete it?) but it showed you are using [connectViewportFrusta](https://www.itwinjs.org/reference/core-frontend/views/connectviewportfrusta/) which only synchronizes the volume of space viewed by each viewport.
You will additionally want to synchronize [ViewDetails.clipVector](https://www.itwinjs.org/reference/core-common/views/viewdetails/clipvector/). Something like:
```
this.disconnect = connectViewports(vp3, (source) => {
  // Clone the frustum (once).
  const pose = source.view.savePose();
  // Get the clip
  const clip = source.details.clipVector;

  // Apply the frustum and clip to each target viewport
  return (_source, target) => {
    const view = target.view.applyPose(pose);
    view.details.clipVector = clip;
  };
});
```

For some reason `applyPose` and `savePose` are marked `@internal`; I'll see about changing that in 3.5 (to be released imminently).
EDIT: See #4737.

**Reply by ovaisa** [reply]
Hi Paul,

Sorry, yes I deleted my post as I wanted to try something and then repost again.  But your timely response pointed me in the direction that helped me implement this capability, so much appreciate your help. 

The feature is now working as expected and clipping is being applied to all viewports.  

However, there is one issue.  When I end clipping (by calling vp.view.setViewClip() with no argument), clipping ends in the current viewport and the entire model is shown.  But other viewports still show the clipped model.  This is because the sync function is not called when I call setViewClip().  As soon as I change the view (by zoom or rotate), the sync function is called and entire model now appears in all viewports. 

I tried calling invalidateScene() after setViewClip() (and tried ViewClipClearTool too) but that also doesn't seem to call the sync function.  Can you please advise how I can force calling the sync function.

Thanks again for your help in this resolving this.

Ovais

**Reply by pmconne** [reply]
> However, there is one issue.

Thanks - I'll fix this up.

**Reply by pmconne** [reply]
A temporary workaround is to call `invalidateController` or `synchWithView(false)` after `setViewClip`.

**Reply by ovaisa** [reply]
I wasn't sure which parameter of ViewChangeOptions to set to false for synchWithView, but invalidateController worked.  Thank you.

Thanks again for providing the code for connectViewport(), this really helped me out!!!  Much appreciated.

Ovais
